pkg/k8s/service_cache.go

```diff
@@ -0,0 +1,384 @@
+
+	Events chan ServiceEvent
+}
+
+// NewServiceCache returns a new ServiceCache
+func NewServiceCache() ServiceCache {
+	return ServiceCache{
+		services:  map[ServiceID]*Service{},
+		endpoints: map[ServiceID]*Endpoints{},
+		ingresses: map[ServiceID]*Service{},
+		Events:    make(chan ServiceEvent, 128),
+	}
+}
+
+// UpdateService parses a Kubernetes service and adds or updates it in the
+// ServiceCache. Returns the ServiceID unless the Kubernetes service could not
+// be parsed and a bool to indicate whether the service was changed in the
+// cache or not.
+func (s *ServiceCache) UpdateService(k8sSvc *v1.Service) ServiceID {
+	svcID, newService := ParseService(k8sSvc)
+	if newService == nil {
+		return svcID
+	}
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	if oldService, ok := s.services[svcID]; ok {
+		if oldService.DeepEquals(newService) {
+			return svcID
+		}
+	}
+
+	s.services[svcID] = newService
+
+	// Check if the corresponding Endpoints resource is already available
+	endpoints, ok := s.endpoints[svcID]
+	if ok {
+		s.Events <- ServiceEvent{
+			Action:     UpdateService,
+			ID:         svcID,
+			Service:    newService,
+			K8sService: k8sSvc,
+			Endpoints:  endpoints,
+		}
+	}
+
+	return svcID
+}
+
+// DeleteService parses a Kubernetes service and removes it from the
+// ServiceCache
+func (s *ServiceCache) DeleteService(k8sSvc *v1.Service) {
+	svcID := ParseServiceID(k8sSvc)
+
+	s.mutex.Lock()
+	oldService, serviceOK := s.services[svcID]
+	endpoints, endpointsOK := s.endpoints[svcID]
+	delete(s.services, svcID)
+	s.mutex.Unlock()
+
+	if serviceOK && endpointsOK {
+		s.Events <- ServiceEvent{
+			Action:     DeleteService,
+			ID:         svcID,
+			Service:    oldService,
+			K8sService: k8sSvc,
+			Endpoints:  endpoints,
+		}
+	}
+}
+
+// ListMissingServices returns a map of all services listed in requiredServices
+// that are not found in the cache.
+func (s *ServiceCache) ListMissingServices(requiredServices versioned.Map) versioned.Map {
+	missing := versioned.NewMap()
+
+	s.mutex.RLock()
+	for uuid, svcObj := range requiredServices {
+		neededSvc := svcObj.Data.(*v1.Service)
+		id := ParseServiceID(neededSvc)
+
+		existingService, ok := s.services[id]
+		if !ok {
+			missing.Add(uuid, svcObj)
+			continue
+		}
+
+		_, newService := ParseService(neededSvc)
+		if !existingService.DeepEquals(newService) {
+			missing.Add(uuid, svcObj)
+		}
+	}
+	s.mutex.RUnlock()
+
+	return missing
+}
+
+// UpdateEndpoints parses a Kubernetes endpoints and adds or updates it in the
+// ServiceCache. Returns the ServiceID unless the Kubernetes endpoints could not
+// be parsed and a bool to indicate whether the endpoints was changed in the
+// cache or not.
+func (s *ServiceCache) UpdateEndpoints(k8sEndpoints *v1.Endpoints) (ServiceID, *Endpoints) {
+	svcID, newEndpoints := ParseEndpoints(k8sEndpoints)
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	if oldEndpoints, ok := s.endpoints[svcID]; ok {
+		if oldEndpoints.DeepEquals(newEndpoints) {
```

Comment:
Create either a GH issue of a function that knows how to perform the deep equal of v1.Endpoints for the fields that actually matter to us.